### PR TITLE
 ~30-40% perf win using 'charCodeAt' in CParser.write()

### PR DESCRIPTION
**Pull request by DLehenbauer**
Uses character codes (i.e., numbers) instead of strings comparison inside of the CParser.write() state machine for a significant perf win (measured on node v8.11.4).

Also includes:
* Benchmark for comparison (```npm i && npm run bench```)
* small set of easy-to-debug tests targeting parser (optional)
* .vscode configuration for debugging tests & benchmark (optional)

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,44 @@
+{
```
[the remark this answers is not in the document]
Comment (evan-king) [reply .vscode/launch.json]: While I respect the advantage of convenience for those who'll use it, I also dislike pushing a particular toolchain, and too many people perceive a) support as direction, and b) configuration clutter as more complicated setup and/or less portable dev workflow.  When I see such files (especially for MS, Java-friendly, or to a lesser degree other monolithic platforms), my snap reaction tends to be "ew, let's find something else." :P

In the absence of an active community to poll, I think I'd prefer to keep the codebase minimalistic and setup guidance toolchain-agnostic (i.e. if you need it, npm install supplied it and package scripts show the way to use it).

File: clarinet.js [hunk 2]
```diff
@@ -70,6 +72,93 @@ else env = window;
   // switcharoo
   S = clarinet.STATE;
 
+  const Char = {
```
Comment (evan-king) [inline clarinet.js]: I have a hard time envisioning a future where letter and number constants have any use.  Perhaps trim this down to just the codes used in this PR?

File: test/parser.spec.js [hunk 3]
```diff
@@ -0,0 +1,178 @@
+"use strict";
```
[the remark this answers is not in the document]
Comment (evan-king) [reply test/parser.spec.js]: I rather favor test suites set up in this way (filthy ES6 classes/OOP notwithstanding ;)), and especially being able to run more focused tests for debugging, so I'm happy to keep it.

File: .vscode/launch.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (evan-king) [reply .vscode/launch.json]: You're welcome to add to `.gitignore` as appropriate.

**Comment by DLehenbauer** [comment]
I was lazy about analyzing the empty string checks.  Looking again, it's pretty easy to convince oneself that the removed empty string checks were previously unreachable.

The normal control flow is to advance ```i++``` until ```charAt(i)``` reads off the end of the string.  This results in ```c``` being assigned to an empty string every time ```write()``` is called.  (i.e., no "strange buffers" required)

When ```c``` is an empty string, the ```if (!c) break;``` near the main while...loop terminates the loop before the switch statement has a chance to dispatch to the S.TRUE*/S.FALSE*/S.NULL* cases.

Since ```c``` is not reassigned before we dispatch to the true/false/null cases, and no one falls through to these cases, we can conclude that ```c``` can not be an empty string when we enter the true/false/null cases (and consequently, the removed empty string checks were indeed unreachable.)

**Comment by evan-king** [comment]
I'm pretty comfortable with that analysis.  However, it does still leave the question of how much performance gain came from removing extraneous conditional checks and not from the use of string literal comparison vs character value constants.

The former makes more sense to me as a cause of sub-optimal performance (especially post-Spectre).  With that in mind, I want to be certain the larger change is justified by a meaningful impact.  Otherwise, significant changes are being introduced for still-unquantified benefit that may not even remain relevant as future engine-level optimizations come down the pipeline, and future maintainers will be as afraid to disrupt or remove performance-related code as I am to introduce it.

**Comment by evan-king** [comment]
Thanks for the thorough PR support.  Approved and merging.

**Pull request merged**